### PR TITLE
Coin Selection Refactor With CAT Coin Selection Refactor 

### PR DESCRIPTION
**Pull request by jack60612**
- [x]  write new coin selection algorithms 
- [x] decompose coin_select function
- [x] Test and use new function with wallet and cat wallet.
- [x] Add integration test

### CODE REVIEW

File: tests/wallet/cat_wallet/test_cat_wallet.py
Comment (Quexington) [inline tests/wallet/cat_wallet/test_cat_wallet.py:773]: I have a PR out to change this boilerplate, it's actually broken right now, if you could update this to work the way it works in that branch that'd be great: https://github.com/Chia-Network/chia-blockchain/pull/9976

File: chia/wallet/coin_selection.py
Comment (mariano54) [inline chia/wallet/coin_selection.py:10]: It's not good practice to pass the larger higher level object here (wallet).

File: tests/wallet/cat_wallet/test_cat_wallet.py
Comment (Quexington) [inline tests/wallet/cat_wallet/test_cat_wallet.py:807]: Isn't the only coin in cat_wallet_2 a coin of 40000?  The old algorithm would still have used it this way right?

File: chia/wallet/coin_selection.py
Comment (mariano54) [reply chia/wallet/coin_selection.py:10]: Instead, we can pass in the required data:
- spendable cons
- constants
- unconfirmed removals

File: tests/wallet/test_wallet.py
Comment (Quexington) [inline tests/wallet/test_wallet.py:806]: Same as above, this boilerplate is actually broken and won't test trusted syncing.

**Review by mariano54 — CHANGES_REQUESTED** [review]

File: tests/wallet/cat_wallet/test_cat_wallet.py
Comment (Quexington) [inline tests/wallet/cat_wallet/test_cat_wallet.py:751]: I'm dubious that we need this test since we've changed the CAT select_coins algorithm to use the shared `select_coins` method.  Is there a reason why we should specifically test this for CATs?

Comment (jack60612) [reply tests/wallet/cat_wallet/test_cat_wallet.py:751]: Nope, ill just remove it.

**Comment by jack60612** [comment]
Closing and opening for CI

File: chia/wallet/cat_wallet/cat_wallet.py
Comment (mariano54) [inline chia/wallet/cat_wallet/cat_wallet.py:409]: type annotation: uint64

**Review by mariano54 — DISMISSED** [review]
LGTM when tests pass

**Review by mariano54 — DISMISSED** [review]

**Comment by mariano54** [comment]
Reopened to trigger test runs

**Comment by jack60612** [comment]
ci

**Review by mariano54 — DISMISSED** [review]

**Review by Quexington — DISMISSED** [review]

File: chia/wallet/wallet.py
Comment (altendky) [inline chia/wallet/wallet.py:3]: isort is going to complain about this now.

```suggestion
from typing import Any, Dict, List, Optional, Set
```

File: tests/wallet/test_coin_selection.py
Comment (altendky) [inline tests/wallet/test_coin_selection.py:8]: isort again

```suggestion
import pytest

from chia.consensus.default_constants import DEFAULT_CONSTANTS
from chia.types.blockchain_format.coin import Coin
from chia.util.hash import std_hash
from chia.util.ints import uint64, uint128
from chia.wallet.coin_selection import check_for_exact_match, find_smallest_coin, knapsack_coin_algorithm
```

**Comment by ftruzzi** [comment]
Hi! Have you thought about exposing coin selection in the RPC API? I've opened a PR for it, and I can rebase it on this branch if needed.

https://github.com/Chia-Network/chia-blockchain/pull/10495

Thanks

**Comment by jack60612** [comment]
> Hi! Have you thought about exposing coin selection in the RPC API? I've opened a PR for it, and I can rebase it on this branch if needed.
> 
> 
> 
> https://github.com/Chia-Network/chia-blockchain/pull/10495
> 
> 
> 
> Thanks
> 
> 
> 
> 

I would wait until the pr is merged but, feel free.

File: chia/wallet/coin_selection.py
Comment (emlowe) [inline chia/wallet/coin_selection.py:33]: Maybe these and others should be Debug instead of Info?

File: chia/wallet/coin_selection.py
Comment (emlowe) [inline chia/wallet/coin_selection.py:90]: The list is sorted, so we could do better than iterating the list if we wanted more performance (binary search anyone - unfortunately since sorted descending, I don't think we can use bisect_left directly). This list of spendable coins could be quite large.

**Comment by emlowe** [comment]
I wonder about adding in support for an "ignore dust" feature - so add in the ability to ignore coins below a certain threshold?

File: chia/wallet/wallet.py
Comment (emlowe) [inline chia/wallet/wallet.py:271]: why are these two asserts here. If these are real errors please Raise something, otherwise just remove them, or if debugging, add debug log entries.
The way I read it is never of these can happen, and if they do it's a serious bug probably worth of a raise of some kind

File: chia/wallet/coin_selection.py
Comment (emlowe) [inline chia/wallet/coin_selection.py:75]: This can't happen, so I think this should raise an Exception since it would be a programming error

File: chia/wallet/coin_selection.py
Comment (emlowe) [inline chia/wallet/coin_selection.py:74]: You could slice valid_spendable_coins by removing all coins too small since they can't possibly be selected - which I think maybe `valid_spendable_coins[:len(smaller_coins)]` would work.

Comment (emlowe) [reply chia/wallet/coin_selection.py:74]: In which case, max_coin_amount would actually be valid_spendable_coins[0].amount, since that is the largest spendable coin and by definition must be > amount. So I think you can rewrite this to not return Optional

File: tests/wallet/test_wallet.py
Comment (emlowe) [inline tests/wallet/test_wallet.py:839]: I'm not sure we need this test - but we need some tests of select_coins() itself to make sure the overall logic is correct. I see tests for the individual functions, but nothing for select_coins(). Some tests with very large lists might be worthwhile - some people will have 10s of thousands of dust coins, we should check that.

File: chia/wallet/coin_selection.py
Comment (jack60612) [reply chia/wallet/coin_selection.py:75]: a lot of these need to be removed as they cant happen, i forgot to remove them while initially debugging.

File: chia/wallet/wallet.py
Comment (jack60612) [reply chia/wallet/wallet.py:271]: yep just needs to be removed.

File: chia/wallet/coin_selection.py
Comment (jack60612) [reply chia/wallet/coin_selection.py:33]: yep

**Comment by jack60612** [comment]
> I wonder about adding in support for an "ignore dust" feature - so add in the ability to ignore coins below a certain threshold?

ill add this in

File: chia/wallet/coin_selection.py
Comment (jack60612) [reply chia/wallet/coin_selection.py:74]: Good point.

File: chia/wallet/coin_selection.py
Comment (jack60612) [reply chia/wallet/coin_selection.py:90]: I am just going to remove this, as it should already be properly sorted.

File: tests/wallet/test_wallet.py
Comment (jack60612) [reply tests/wallet/test_wallet.py:844]: Good point, after i separated the function i forgot to separate this.

File: chia/wallet/coin_selection.py
Comment (mariano54) [reply chia/wallet/coin_selection.py]: No, they shouldn't be removed if they can't happen. That is the use cases for asserts

Comment (jack60612) [reply chia/wallet/coin_selection.py:85]: Re added

File: chia/wallet/wallet.py
Comment (jack60612) [reply chia/wallet/wallet.py]: Re added

**Review by mariano54 — DISMISSED** [review]

**Review by mariano54 — DISMISSED** [review]

**Comment by mariano54** [comment]
Need this first:
https://github.com/Chia-Network/chia-blockchain/pull/10967

File: chia/wallet/coin_selection.py
Comment (emlowe) [inline chia/wallet/coin_selection.py:93]: Suggest more concise string `raise ValueError(f"Transaction of {amount} would use more than 500 coins. Try sending a smaller amount")

File: chia/wallet/coin_selection.py
Comment (emlowe) [inline chia/wallet/coin_selection.py:121]: `ValueError(f"Unable to select coins for transaction. Try sending a smaller amount")

File: chia/wallet/coin_selection.py
Comment (emlowe) [inline chia/wallet/coin_selection.py:56]: `ValueError("Transaction for {amount} is greater than spendable balance of {sum_spendable_coins}. There may be other transactions pending")`

**Comment by emlowe** [comment]
I made a few suggestions on error strings - I don't think we should mention anything about dust in the error string. (I'd probably prefer returning some error code instead, or a custom exception, but we don't historically do that)

**Review by mariano54 — APPROVED** [review]

**Comment by emlowe** [comment]
I was looking at `get_max_send_amount` in both the standard wallet and the cat wallet, and I notice that it doesn't take into account the coin selection algorithm at all. It didn't before (so I think it would give wrong answers often) and it doesn't in this PR either (and so likely still gives wrong answers) - This is trying to actually determine the maximum number of coins that we want to fit in a transaction, but since it doesn't follow the coin selection algorithm it won't pick the correct coins for this calculation.

I'm not sure the best way to fix this function but it seems rather incorrect right now.

**Comment by emlowe** [comment]
> I was looking at `get_max_send_amount` in both the standard wallet and the cat wallet, and I notice that it doesn't take into account the coin selection algorithm at all. EDIT: I was mistaken - it does this `spendable.sort(reverse=True, key=lambda record: record.coin.amount)` , so it does sort by biggest first like the old coin algorithm. However, it certainly doesn't use the new algorithm which will pick very different coins.
> 
> I'm not sure the best way to fix this function but it seems rather incorrect in this PR because it will choose different coins to calculate this max vs what the coin selection algo will do.

Oh, maybe this is ok - this function is actually trying to determine the maximum amount you could send if you choose all the biggest coins with a max number of coins set heuristically based on the standard wallet TX and some arbitrary scaling of max clvm cost. This does mean though that it's unlikely the code that hits coin limits in the new coin selection actually happen in the wild, since this function is check first.

So I think it's fine to merge in this change without further changed here.

**Review by emlowe — APPROVED** [review]

**Pull request merged**
